contracts/ReferralHandler.sol

```diff
@@ -0,0 +1,193 @@
+        require(
+            !addressExists[epochNonce][msg.sender],
+            "Rebate already claimed"
+        );
+        uint256 amountLeftToClaim = maxAmountToClaim - amountClaimed;
+        require(amountLeftToClaim > 0, "All rebates have been Claimed");
+        require(_verifyProof(proof, msg.sender, amount), "Invalid Proof");
+        addressExists[epochNonce][msg.sender] = true;
+        uint256 ClaimableAmount = amount < amountLeftToClaim
+            ? amount
+            : amountLeftToClaim;
+        amountClaimed += ClaimableAmount;
+        // Follows checks effects pattern. So there should be no re-entrancy exploit
+        rewardToken.safeTransfer(msg.sender, ClaimableAmount);
+        emit RewardPaid(msg.sender, ClaimableAmount);
+    }
+
+    /// @inheritdoc IRebateHandler
+    function reclaimUnusedReward(address account) external onlyOwner {
+        require(
+            (block.timestamp - lastUpdated >= timeToReclaimRewards),
+            "time less than timeToReclaimRewards"
+        );
+        rewardToken.safeTransfer(account, rewardToken.balanceOf(address(this)));
+    }
+
+    /// @notice allows admin to update merkleRootAdmin variable
+    /// @param _merkleRootAdmin new merkleRootAdmin value.
+    function updateMerkleRootAdmin(
+        address _merkleRootAdmin
+    ) external onlyOwner {
+        require(
+            _merkleRootAdmin != address(0),
+            "_merkleRootAdmin can't be zero"
+        );
+        require(
+            _merkleRootAdmin != owner(),
+            "_merkleRootAdmin can't be same as Owner"
+        );
+        merkleRootAdmin = _merkleRootAdmin;
+    }
+
+    /// @notice allows admin to update timeElapsedForUpdate variable
+    /// @param _timeElapsedForUpdate new timeElapsedForUpdate value.
+    /// @dev don't allow to update unless previous time has passed
+    function updateTimeElapsedForUpdate(
+        uint256 _timeElapsedForUpdate
+    ) external onlyOwner {
+        require(_timeElapsedForUpdate != 0, "_timeElapsedForUpdate can't be 0");
+        require(
+            (block.timestamp - lastUpdatedTimeElapsed >= timeElapsedForUpdate),
+            "Cannot update before timeElapsedForUpdate ends"
+        );
+        lastUpdatedTimeElapsed = block.timestamp;
+        timeElapsedForUpdate = _timeElapsedForUpdate;
+    }
+
+    /// @notice allows admin to update timeToReclaimRewards variable
+    /// @param _timeToReclaimRewards new timeToReclaimRewards value.
+    /// TODO: admin can manipulate this, add a check on times
+    function updateTimeToReclaimRewards(
+        uint256 _timeToReclaimRewards
+    ) external onlyOwner {
+        require(_timeToReclaimRewards != 0, "_timeToReclaimRewards can't be 0");
+        require(
+            _timeToReclaimRewards > timeElapsedForUpdate,
+            "value less than timeElapsedForUpdate"
+        );
+        timeToReclaimRewards = _timeToReclaimRewards;
+    }
+
+    function _resetAddressExists() internal {
```

Comment:
This is a highly misleading function. This doesn't reset anything. Get rid of this function and increment the nonce directly in the `updateMerkleRoot` function.